core/coreapi/unixfs.go

```diff
@@ -53,31 +51,27 @@ func (api *UnixfsAPI) Add(ctx context.Context, files files.Node, opts ...options
 		return nil, filestore.ErrFilestoreNotEnabled
 	}
 
-	addblockstore := api.blockstore
-	if !(settings.FsCache || settings.NoCopy) {
-		addblockstore = bstore.NewGCBlockstore(api.baseBlocks, api.blockstore)
-	}
-	exch := api.exchange
-	pinning := api.pinning
+	var (
+		pinning       pin.Pinner
+		dserv         ipld.DAGService
+		addblockstore bstore.GCBlockstore
+	)
 
-	if settings.OnlyHash {
-		nilnode, err := core.NewNode(ctx, &core.BuildCfg{
-			//TODO: need this to be true or all files
-			// hashed will be stored in memory!
-			NilRepo: true,
-		})
-		if err != nil {
-			return nil, err
+	if !settings.OnlyHash {
+		addblockstore = api.blockstore
+		if !(settings.FsCache || settings.NoCopy) {
+			// XXX: Why do we need this?
+			addblockstore = bstore.NewGCBlockstore(api.baseBlocks, addblockstore)
+		}
+		if settings.Pin {
+			pinning = api.pinning
 		}
-		addblockstore = nilnode.Blockstore
-		exch = nilnode.Exchange
-		pinning = nilnode.Pinning
-	}
 
-	bserv := blockservice.New(addblockstore, exch) // hash security 001
-	dserv := dag.NewDAGService(bserv)
+		bserv := blockservice.New(addblockstore, api.exchange) // hash security 001
+		dserv = dag.NewDAGService(bserv)
+	}
 
-	fileAdder, err := coreunix.NewAdder(ctx, pinning, addblockstore, dserv)
+	fileAdder, err := coreunix.NewAdder(pinning, addblockstore, dserv)
```

Comment:
TL;DR:

1. We can now pass a nil dagservice to "only hash". I needed to move this functionality down into the adder because I removed the SetMfsRoot function.
2. Instead of setting `adder.Pin`, we now just pass a nil pinner to disable pinning. Really, I only _needed_ 1 (see below) but I figured I'd make these consistent.

Alternatively, I could have added an `OnlyHash` option to the file adder.

Background:

1. I wanted to move away from MFS for adding single files to get away from all the directory wrapping issues. However, the old adder kind of assumed MFS everywhere. It was technically possible to just not use it but I was worried about leaving the adder in an inconsistent state.
2. The old adder wasn't safely reusable (as far as I can tell).

So I refactored it into:

1. `Adder`: A reusable, immutable "adder" to carry all the add options, dag/pin/gc services, etc.
2. `adderJob`: job for adding files/symlinks (embeds the `Adder`).
3. `dirAdderJob`: job for adding directories (with mfs support) (embeds `adderJob`).

How does that relate to the changes above? Well, `Adder` no longer stores the MFS root (because adder is immutable/reusable). That meant `SetMfsRoot` had to go.

---

Really, we probably shouldn't be using MFS for importing files _at all_. I'm not sure why we are but we can fix that later.